numba/targets/gdb_hook.py

```diff
@@ -150,27 +169,63 @@ def init_gdb_codegen(cgctx, builder, signature, args,
         msg = "Internal error: `fork` failed."
         cgctx.call_conv.return_user_exc(builder, RuntimeError, (msg,))
 
+    # call pipe to set up fds
+    pipefds_ptr = builder.gep(pipefds, [zero_i32t], inbounds=True)
+    stat = builder.call(pipe, (pipefds_ptr,))
+    pipe_failed = builder.icmp_signed('==', stat, int32_t(-1))
+    with builder.if_then(pipe_failed, likely=False):
+        msg = "Internal error: `pipe` failed."
+        cgctx.call_conv.return_user_exc(builder, RuntimeError, (msg,))
+
+    # pointers to the read and write end FDs
+    read_end = builder.gep(pipefds, [zero_i32t], inbounds=True)
+    write_end = builder.gep(pipefds, [one_i32t], inbounds=True)
+
+    # this writes a command like "close(write_end_fd)" into the text buffer that
+    # will be one of the const char * that goes into the execl call made to
+    # gdb with view of a command like `-ex "close(write_end_fd)"` being executed
+    stat = builder.call(snprintf, (callclose_str, intp_t(20), callclosefmt,
+                                   builder.load(write_end)))
+    invalid_write = builder.icmp_signed('>', stat, int32_t(20))
+    with builder.if_then(invalid_write, likely=False):
+        msg = "Internal error: `snprintf` buffer would have overflowed."
+        cgctx.call_conv.return_user_exc(builder, RuntimeError, (msg,))
+
     is_child = builder.icmp_signed('==', child_pid, zero_i32t)
     with builder.if_else(is_child) as (then, orelse):
         with then:
             # is child
+
+            # first close the fork-inherited pipe ends, doesn't matter if they
+            # return something
+            builder.call(close, (builder.load(write_end),))
```

Comment:
This code is once removed from what's happening, so I could be missing it, but I think there's two processes: the parent that needs to wait by reading from a pipe(2); the child that will replace itself with gdb(1) that will attach to a process and then close(2) the write end of the pipe, signalling the parent. But here the child appears to be closing the write end before the execve(2). gdb's later close() will either return EBADF or it will be a valid FD that gdb has opened for its own purposes. One way to quickly observe this might be `strace -fe pipe,read,close a.out` and follow the path from the pipe(). Or, if one's left at a gdb prompt, don't close() from gdb and then examine _/proc/$gdbpid/fd_.

Reply:
Thanks for looking at this. I think the child can close both the read and write ends of its own fork(2)'d FDs as they are of no use. The image replacement with gdb(1) and then subsequent attach will give the child access to the FDs of the parent. The parent is blocking on the read(2) and the gdb(1) attachment will close(2) the write end of the parent FD on start up `-ex close(parent_write_fd)` as a signal to unblock as it has access to the parent FDs by virtue of the attach. I think there may be a FD leak but it will be from the parent read end of the pipe after fork (guess this is to what this https://github.com/numba/numba/pull/3957/files#r273386107 refers)?

Reply:
For reference, Python source, LLVM IR, asm, and runtime log: https://gist.github.com/stuartarchibald/2ff1d5b93ef99187c1a9b67d46974796

Reply:
But the image replacement by gdb has gdb inherit the open FDs of the process doing the exec, the child, it doesn't get a fresh copy of the parent's FDs, and so closing both pipe ends in the child before the exec means the parent is signalled early.  The number of the file descriptor can still be used in gdb, but it's either already closed, EBADF, or that number has been allocated by the kernel due to another open(2), etc.

I think this bash shows the problem that I think exists. It works fine as it is; the parent only continues after the first two-second sleep and strace shows the successful `close(1)` after the marker `close(42)`. Uncommenting the extra `exec` has the `read` exit early and `close(1)` suffers EBADF.

    #! /bin/bash

    child() {
        exec 0<&- # Close stdin.
        # Uncomment the next line to show the bug.
        # exec 1>&- # Close stdout.
        sleep 2 # Parent should wait for this to finish.
        exec strace -e close sh -c '
                exec 42<&- 1>&-;
                sleep 2;
                echo child exit >&2
            ' # Have gdb close closed stdout.
    }

    coproc child

    echo `date` starting
    read <&${COPROC[0]}
    echo `date` read exited $?
    wait
    echo `date` child exited

Reply:
Ah, I'm wrong. Discussion with @stuartarchibald has reminded me that the process gdb is attaching to is the parent that does the fork thus gdb's call to close() is altering the parent's open FDs and the child needs to have already closed its FD for the write end of the pipe so all references to the write end are closed.

Reply:
Thanks for looking at this again @RalphCorderoy. In the present state (3231816) this gdb(1) session shows that the read end (fd=5) is open as a pipe when the breakpoint is hit by the parent and gdb(1) itself has fd's 3-10 open as pipes.

```
0x00007f4f06de76e0 in __read_nocancel () at ../sysdeps/unix/syscall-template.S:81
81      T_PSEUDO (SYSCALL_SYMBOL, SYSCALL_NAME, SYSCALL_NARGS)
$1 = 0
Breakpoint 1 at 0x7f4eefce88f0: file numba/_helperlib.c, line 1133.
Continuing.

Breakpoint 1, numba_gdb_breakpoint () at numba/_helperlib.c:1133
warning: Source file is more recent than executable.
1133    }

(gdb) bt
#0  numba_gdb_breakpoint () at numba/_helperlib.c:1133
#1  0x00007f4effa0947e in numba::targets::gdb_hook::hook_gdb::$3clocals$3e::impl$242(Tuple<Literal$5bstr$5d$28$2dex$29, Literal$5bstr$5d$28info$20proc$29>) ()
#2  0x00007f4efe7c5088 in __main__::foo$241 () at test_gdb1.py:5
#3  0x00007f4efe7c5211 in cpython::__main__::foo$241 ()
#4  0x00007f4eef8976b4 in call_cfunc ()
   from <src_path>/numba/_dispatcher.cpython-37m-x86_64-linux-gnu.so
#5  0x00007f4eef897af9 in compile_and_invoke ()
   from <src_path>/numba/_dispatcher.cpython-37m-x86_64-linux-gnu.so
#6  0x00007f4eef897c81 in Dispatcher_call ()
   from <src_path>/numba/_dispatcher.cpython-37m-x86_64-linux-gnu.so
#7  0x0000560302a2f65b in _PyObject_FastCallKeywords () at /tmp/build/80754af9/python_1544812943833/work/Objects/call.c:199
#8  0x0000560302a8ad96 in call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1544812943833/work/Python/ceval.c:4605
#9  _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1544812943833/work/Python/ceval.c:3124
#10 0x00005603029cbf19 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1544812943833/work/Python/ceval.c:3930
#11 0x00005603029ccdd4 in PyEval_EvalCodeEx () at /tmp/build/80754af9/python_1544812943833/work/Python/ceval.c:3959
#12 0x00005603029ccdfc in PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>)
    at /tmp/build/80754af9/python_1544812943833/work/Python/ceval.c:524
#13 0x0000560302ae59e4 in run_mod () at /tmp/build/80754af9/python_1544812943833/work/Python/pythonrun.c:1035
#14 0x0000560302aefbd1 in PyRun_FileExFlags () at /tmp/build/80754af9/python_1544812943833/work/Python/pythonrun.c:988
#15 0x0000560302aefdc3 in PyRun_SimpleFileExFlags () at /tmp/build/80754af9/python_1544812943833/work/Python/pythonrun.c:429
#16 0x0000560302af0edb in pymain_run_file (p_cf=0x7fff2e678680, filename=0x5603030602e0 L"test_gdb1.py", fp=0x5603030dac80)
    at /tmp/build/80754af9/python_1544812943833/work/Modules/main.c:427
#17 pymain_run_filename (cf=0x7fff2e678680, pymain=0x7fff2e678790)
    at /tmp/build/80754af9/python_1544812943833/work/Modules/main.c:1537
#18 pymain_run_python (pymain=0x7fff2e678790) at /tmp/build/80754af9/python_1544812943833/work/Modules/main.c:2626
#19 pymain_main () at /tmp/build/80754af9/python_1544812943833/work/Modules/main.c:2787
#20 0x0000560302af0fbc in _Py_UnixMain () at /tmp/build/80754af9/python_1544812943833/work/Modules/main.c:2822
#21 0x00007f4f06a2e3d5 in __libc_start_main (main=0x5603029ac830 <main>, argc=2, argv=0x7fff2e6788e8, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fff2e6788d8) at ../csu/libc-start.c:266
#22 0x0000560302a95ed2 in _start () at ../sysdeps/x86_64/elf/start.S:103
(gdb) info proc
process 23424
warning: target file /proc/23424/cmdline contained unexpected null characters
cmdline = 'python'
cwd = '<src_path>'
exe = '<py_path>/bin/python3.7'
(gdb) shell ls -l /proc/23424/fd
total 0
lrwx------. 1 <user> <user> 64 Apr  9 12:53 0 -> /dev/pts/1
lrwx------. 1 <user> <user> 64 Apr  9 12:53 1 -> /dev/pts/1
lrwx------. 1 <user> <user> 64 Apr  9 12:53 2 -> /dev/pts/1
lr-x------. 1 <user> <user> 64 Apr  9 12:53 3 -> /dev/urandom
lrwx------. 1 <user> <user> 64 Apr  9 12:53 4 -> /tmp/ffiVqVBC2 (deleted)
lr-x------. 1 <user> <user> 64 Apr  9 12:53 5 -> pipe:[431497]
(gdb) shell ps aux|grep 'gdb attach'
<user>      23429 11.7  3.3 487952 271544 pts/1   S+   12:53   0:04 /usr/bin/gdb attach 23424 -ex call close(6
<user>      23437  0.0  0.0 113184  1420 pts/1    S+   12:53   0:00 bash -c ps aux|grep 'gdb attach'
<user>      23439  0.0  0.0 112716   964 pts/1    S+   12:53   0:00 grep gdb attach
(gdb) shell ls -l /proc/23429/fd|grep pipe
l-wx------. 1 <user> <user> 64 Apr  9 12:54 10 -> pipe:[428547]
lr-x------. 1 <user> <user> 64 Apr  9 12:53 3 -> pipe:[428544]
l-wx------. 1 <user> <user> 64 Apr  9 12:54 4 -> pipe:[428544]
lr-x------. 1 <user> <user> 64 Apr  9 12:54 5 -> pipe:[428545]
l-wx------. 1 <user> <user> 64 Apr  9 12:54 6 -> pipe:[428545]
lr-x------. 1 <user> <user> 64 Apr  9 12:54 7 -> pipe:[428546]
l-wx------. 1 <user> <user> 64 Apr  9 12:54 8 -> pipe:[428546]
lr-x------. 1 <user> <user> 64 Apr  9 12:54 9 -> pipe:[428547]

```

Fixing the source (716cd2e4f) so that the `pipe` call happens before the `fork` (which was the original intention but an ommission, still worked because nothing else had fd's so lowest were assigned symmetrically between parent and child), and then adding a `close(2)` on the read end of the pipe immediately after the read(2) call is unblocked gives the following:

```
0x00007f3e3939d6e0 in __read_nocancel () at ../sysdeps/unix/syscall-template.S:81
81      T_PSEUDO (SYSCALL_SYMBOL, SYSCALL_NAME, SYSCALL_NARGS)
$1 = 0
Breakpoint 1 at 0x7f3e2229e8f0: file numba/_helperlib.c, line 1133.
Continuing.

Breakpoint 1, numba_gdb_breakpoint () at numba/_helperlib.c:1133
warning: Source file is more recent than executable.
1133    }

(gdb) bt
#0  numba_gdb_breakpoint () at numba/_helperlib.c:1133
#1  0x00007f3e31fbf498 in numba::targets::gdb_hook::hook_gdb::$3clocals$3e::impl$242(Tuple<Literal$5bstr$5d$28$2dex$29, Literal$5bstr$5d$28info$20proc$29>) ()
#2  0x00007f3e30d7b088 in __main__::foo$241 () at test_gdb1.py:5
#3  0x00007f3e30d7b211 in cpython::__main__::foo$241 ()
#4  0x00007f3e21e4d6b4 in call_cfunc ()
   from <src_path>/numba/_dispatcher.cpython-37m-x86_64-linux-gnu.so
#5  0x00007f3e21e4daf9 in compile_and_invoke ()
   from <src_path>/numba/_dispatcher.cpython-37m-x86_64-linux-gnu.so
#6  0x00007f3e21e4dc81 in Dispatcher_call ()
   from <src_path>/numba/_dispatcher.cpython-37m-x86_64-linux-gnu.so
#7  0x00005620c072565b in _PyObject_FastCallKeywords () at /tmp/build/80754af9/python_1544812943833/work/Objects/call.c:199
#8  0x00005620c0780d96 in call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1544812943833/work/Python/ceval.c:4605
#9  _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1544812943833/work/Python/ceval.c:3124
#10 0x00005620c06c1f19 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1544812943833/work/Python/ceval.c:3930
#11 0x00005620c06c2dd4 in PyEval_EvalCodeEx () at /tmp/build/80754af9/python_1544812943833/work/Python/ceval.c:3959
#12 0x00005620c06c2dfc in PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>)
    at /tmp/build/80754af9/python_1544812943833/work/Python/ceval.c:524
#13 0x00005620c07db9e4 in run_mod () at /tmp/build/80754af9/python_1544812943833/work/Python/pythonrun.c:1035
#14 0x00005620c07e5bd1 in PyRun_FileExFlags () at /tmp/build/80754af9/python_1544812943833/work/Python/pythonrun.c:988
#15 0x00005620c07e5dc3 in PyRun_SimpleFileExFlags () at /tmp/build/80754af9/python_1544812943833/work/Python/pythonrun.c:429
#16 0x00005620c07e6edb in pymain_run_file (p_cf=0x7ffd6409f820, filename=0x5620c26552e0 L"test_gdb1.py", fp=0x5620c26cfc80)
    at /tmp/build/80754af9/python_1544812943833/work/Modules/main.c:427
#17 pymain_run_filename (cf=0x7ffd6409f820, pymain=0x7ffd6409f930)
    at /tmp/build/80754af9/python_1544812943833/work/Modules/main.c:1537
#18 pymain_run_python (pymain=0x7ffd6409f930) at /tmp/build/80754af9/python_1544812943833/work/Modules/main.c:2626
#19 pymain_main () at /tmp/build/80754af9/python_1544812943833/work/Modules/main.c:2787
#20 0x00005620c07e6fbc in _Py_UnixMain () at /tmp/build/80754af9/python_1544812943833/work/Modules/main.c:2822
#21 0x00007f3e38fe43d5 in __libc_start_main (main=0x5620c06a2830 <main>, argc=2, argv=0x7ffd6409fa88, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffd6409fa78) at ../csu/libc-start.c:266
#22 0x00005620c078bed2 in _start () at ../sysdeps/x86_64/elf/start.S:103
(gdb) info proc
process 23532
warning: target file /proc/23532/cmdline contained unexpected null characters
cmdline = 'python'
cwd = '<src_path>'
exe = '<py_path>/bin/python3.7'
(gdb) shell ls -l /proc/23532/fd
total 0
lrwx------. 1 <user> <user> 64 Apr  9 12:55 0 -> /dev/pts/1
lrwx------. 1 <user> <user> 64 Apr  9 12:55 1 -> /dev/pts/1
lrwx------. 1 <user> <user> 64 Apr  9 12:55 2 -> /dev/pts/1
lr-x------. 1 <user> <user> 64 Apr  9 12:55 3 -> /dev/urandom
lrwx------. 1 <user> <user> 64 Apr  9 12:55 4 -> /tmp/ffiC4haNz (deleted)
(gdb) shell ps aux|grep 'gdb attach'
<user>      23537  9.3  3.3 487952 271288 pts/1   S+   12:54   0:04 /usr/bin/gdb attach 23532 -ex call close(6
<user>      23551  0.0  0.0 113184  1416 pts/1    S+   12:55   0:00 bash -c ps aux|grep 'gdb attach'
<user>      23553  0.0  0.0 112716   964 pts/1    S+   12:55   0:00 grep gdb attach
(gdb) shell ls -l /proc/23537/fd|grep pipe
l-wx------. 1 <user> <user> 64 Apr  9 12:55 10 -> pipe:[429757]
lr-x------. 1 <user> <user> 64 Apr  9 12:54 3 -> pipe:[429754]
l-wx------. 1 <user> <user> 64 Apr  9 12:55 4 -> pipe:[429754]
lr-x------. 1 <user> <user> 64 Apr  9 12:55 5 -> pipe:[429755]
l-wx------. 1 <user> <user> 64 Apr  9 12:55 6 -> pipe:[429755]
lr-x------. 1 <user> <user> 64 Apr  9 12:55 7 -> pipe:[429756]
l-wx------. 1 <user> <user> 64 Apr  9 12:55 8 -> pipe:[429756]
lr-x------. 1 <user> <user> 64 Apr  9 12:55 9 -> pipe:[429757]
```
which shows that there are now no fd's open as pipes in the parent and that in the child gdb(1) again as fd's 3-10 open as pipes. I believe the leak is fixed.